metadataproxy/roles.py

```diff
@@ -157,117 +157,129 @@ def find_container(ip):
-
-
 @log_exec_time
-def get_role_name_from_ip(ip, stripped=True):
+def get_role_params_from_ip(ip, requested_role=None):
+    params = dict(name=None, account_id=None, external_id=None, session_name=None)
+    role_name = None
     if app.config['ROLE_MAPPING_FILE']:
-        return ROLE_MAPPINGS.get(ip, app.config['DEFAULT_ROLE'])
-    container = find_container(ip)
-    if container:
-        env = container['Config']['Env']
-        for e in env:
-            key, val = e.split('=', 1)
-            if key == 'IAM_ROLE':
-                if val.startswith('arn:aws'):
-                    m = RE_IAM_ARN.match(val)
-                    val = '{0}@{1}'.format(m.group(2), m.group(1))
-                if stripped:
-                    return val.split('@')[0]
-                else:
-                    return val
-        msg = "Couldn't find IAM_ROLE variable. Returning DEFAULT_ROLE: {0}"
-        log.debug(msg.format(app.config['DEFAULT_ROLE']))
-        if stripped:
-            return app.config['DEFAULT_ROLE'].split('@')[0]
+        role = ROLE_MAPPINGS.get(ip, app.config['DEFAULT_ROLE'])
+        if isinstance(role, dict):
+            params.update(role)
         else:
-            return app.config['DEFAULT_ROLE']
+            role_name = role
     else:
-        return None
+        container = find_container(ip)
+        if container:
+            env = container['Config']['Env'] or []
+            # Look up IAM_ROLE and IAM_EXTERNAL_ID values from environment
+            for e in env:
+                key, val = e.split('=', 1)
+                if key == 'IAM_ROLE':
+                    if val.startswith('arn:aws'):
+                        m = RE_IAM_ARN.match(val)
+                        val = '{0}@{1}'.format(m.group(2), m.group(1))
+                    role_name = val
+                elif key == 'IAM_EXTERNAL_ID':
+                    params['external_id'] = val
+            if not role_name:
+                msg = "Couldn't find IAM_ROLE variable. Returning DEFAULT_ROLE: {0}"
+                log.debug(msg.format(app.config['DEFAULT_ROLE']))
+                role_name = app.config['DEFAULT_ROLE']
+
+            # Optionally, look up role session name from environment or labels
+            if app.config['ROLE_SESSION_KEY']:
+                skey = app.config['ROLE_SESSION_KEY']
+                sval = None
+                if skey.startswith('Env:'):
+                    skey = skey[4:]
+                    for e in env:
+                        key, val = e.split('=', 1)
+                        if skey == key:
+                            sval = val
+                elif skey.startswith('Labels:'):
+                    skey = skey[7:]
+                    if container['Config']['Labels'] and skey in container['Config']['Labels']:
+                        sval = container['Config']['Label'][skey]
+                if sval:
+                    params['session_name'] = re.sub(r'[^a-zA-Z0-9\=\,\.\_\-]', '_', sval)[:64]
```

Comment:
Can you add some code docs here? What's the regex do? Why's this limited to 64 char?